enclave_build/docs/cache.md

```diff
@@ -0,0 +1,127 @@
+
+#### Layers
+
+Layers are serialized filesystem changes that go on top of each other, starting from the base filesystem we create the image with. The file will be binary content that, from Nitro CLI's point of view, just has to obey the content digest in the manifest.
+
+## Operations
+
+The operations offered by the `cache` module are as follows:
+
+### Create
+
+When building our first image, the cache will probably not exist. This operation will create, prior to storing an image, an empty cache. It will contain the cache root folder, an index file, a blobs folder and the oci-layout.
+
+### Store
+
+This is the operation which adds an image to the cache. It will be called on an empty cache or after a cache miss when searching for the image. After pulling the image and extracting its details, its manifest entry will be added to the index and the blobs will be written, each to their file.
+
+### Validate
+
+After we find an image in the index or the in-memory cache, we want to validate its structure before using the data. The validation starts from the index file. It should contain our image's reference and the digest of the manifest. If fields are missing or the JSON structure does not comply, an error will be returned.
+
+The validation continues with the manifest. As we did with the index, we validate the structure and get the digests for the config and blobs to later check them as well. We also check that the content of the manifest matches its digest from the file name.
+
+The config validation simply checks that a file with its content digest exists and the content is UTF-8 formatted. The file name is also validated against the content hash.
+
+Having a list of the layers' digest from the manifest, we go through each of them to get access to the blob files. By hashing each of the files and comparing the result with the initial digest, we validate the layers as well.
+
+### Fetch details
+
+This operation will return the details of an image (image name, config digest and the config itself). These details are saved in the memory of the `CacheManager` and will hold the information needed to build the enclave image. This operation is called after validating the cache.
+
+### Get root folder
+
+So we can get access to the cache contents outside the `CacheManager` (in our case for `linuxkit`), we want to be able to return the location of the cache files. It will return the root folder of the cache, as decided by the values of the environment variables `XDG_DATA_HOME` or `HOME`.
+
+## Workflow
+
+To see how these operation fit with each other and the structure as well, let's explore the common workflow:
+
+1. Given an image name we want to build an EIF
+2. Check if the cache exists. If not, pull the image and store it to the new cache.
+3. Check if the image is in the cache (in-memory). If not, pull the image store add it to the cache.
+4. Validate the image's files in the cache. If it fails, pull the image and overwrite cache.
```

Comment:
```suggestion
4. Validate the image's files in the cache. If the validation fails, pull the image and overwrite cache.
```

Reply:
Overwrite cache with what? Do we remove the image or what are we doing?

Reply:
Write the new content of the invalid files and truncate the file.